alf/algorithms/sac_algorithm.py

```diff
@@ -224,63 +219,54 @@ def _actor_train_step(self, exp: Experience, state: SacActorState,
-                                            self._dqda_clipping)
+                    dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                       self._dqda_clipping)
                 loss = 0.5 * losses.element_wise_squared_loss(
-                    tf.stop_gradient(dqda + action), action)
-                loss = tf.reduce_sum(
-                    loss, axis=list(range(1, len(loss.shape))))
+                    (dqda + action).detach(), action)
+                loss = loss.view(loss.size(0), -1).mean(1)
                 return loss
 
-            actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
-            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
+            actor_loss = nest.map_structure(actor_loss_fn, dqda, action)
+            alpha = torch.exp(self._log_alpha).detach()
             actor_loss += alpha * log_pi
         else:
             critic1, critic1_state = self._critic_network1(
-                exp.observation,
-                step_type=exp.step_type,
-                network_state=state.critic1)
+                exp.observation, state=state.critic1)
 
             critic2, critic2_state = self._critic_network2(
-                exp.observation,
-                step_type=exp.step_type,
-                network_state=state.critic2)
+                exp.observation, state=state.critic2)
 
-            assert isinstance(action_distribution,
-                              tfp.distributions.Categorical), (
-                                  "Only `tfp.distributions.Categorical` " +
-                                  "was supported, received:" + str(
-                                      type(action_distribution)))
+            assert isinstance(
+                action_distribution, td.categorical.Categorical) or \
+                isinstance(action_distribution, td.independent.Independent) and \
+                isinstance(action_distribution.base_dist,
+                                td.categorical.Categorical),  \
+                 ("Only `Categorical` " + "was supported, received:" + str(
+                        type(action_distribution)))
 
-            action_probs = action_distribution.probs
-            log_action_probs = tf.math.log(action_probs + 1e-8)
+            log_action_probs = action_distribution.base_dist.logits.squeeze(1)
 
-            target_q_value = tf.stop_gradient(tf.minimum(critic1, critic2))
-            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
-            actor_loss = tf.reduce_mean(
-                action_probs * (alpha * log_action_probs - target_q_value),
-                axis=-1)
+            target_q_value = torch.min(critic1, critic2).detach()
+            alpha = torch.exp(self._log_alpha)
+            actor_loss = torch.exp(log_action_probs) * (
+                alpha.detach() * log_action_probs - target_q_value)
+            actor_loss = actor_loss.view(actor_loss.size(0), -1).mean(1)
```

Comment:
`view()` may not always work. Either avoid it or use reshape:
```python
actor_loss.mean(list(range(1, actor_loss.ndim)))
```

Reply:
Changed according to suggestion.